src/katsdpmodels/primary_beam.py

```diff
@@ -0,0 +1,256 @@
+    (the absolute signs don't matter but the signs relative to each other do
+    for cross-hand terms).
+
+    An instance of this class describes either a single antenna or an average
+    of multiple antennas; to do antenna-specific modelling you will need a
+    separate instance of this class per antenna.
+    """
+
+    model_type: ClassVar[Literal['primary_beam']] = 'primary_beam'
+
+    def resolution(self, frequency: u.Quantity) -> float:
+        """Approximate spatial resolution of the model, in units of projected coordinates.
+
+        Sampling a grid at significantly higher resolution than this will have
+        diminishing returns compared to sampling at this resolution and
+        interpolating.
+        """
+        raise NotImplementedError()
+
+    @property
+    def frequency_range(self) -> Tuple[u.Quantity, u.Quantity]:
+        """Minimum and maximum frequency covered by the model."""
+        raise NotImplementedError()
+
+    @property
+    def frequency_resolution(self) -> u.Quantity:
+        """Approximate frequency resolution of the model.
+
+        Sampling at significantly higher spectral resolution than this will
+        have diminishing returns compared to sampling at this resolution and
+        interpolating.
+        """
+        raise NotImplementedError()
+
+    def min_radius(self, frequency: u.Quantity) -> float:
+        """Maximum distance from the pointing centre at which model has full coverage."""
+        raise NotImplementedError()
+
+    def max_radius(self, frequency: u.Quantity) -> float:
+        """Maximum distance from the pointing centre at which model has any coverage."""
+        raise NotImplementedError()
+
+    @property
+    def is_circular(self) -> bool:
+        """Whether this model is circularly-symmetric about the pointing centre."""
+        raise NotImplementedError()
+
+    @property
+    def is_unpolarized(self) -> bool:
+        """Whether this model ignores polarization.
+
+        If true, it is guaranteed that the Jones matrices describing the beam are
+        scaled identity matrices.
+        """
+        raise NotImplementedError()
+
+    @property
+    def parameters(self) -> Sequence[Parameter]:
+        """Parameters of the model."""
+        raise NotImplementedError()
+
+    def sample(self, l: ArrayLike, m: ArrayLike, frequency: u.Quantity,   # noqa: E741
+               frame: Union[AltAzFrame, RADecFrame],
+               output_type: OutputType,
+               parameters: Dict[str, u.Quantity] = {}) -> np.ndarray:
+        """Sample the primary.
+
+        A sample is returned for each combination of a position (given by `l`,
+        `m`) with a frequency. The dimensions of the output will be first those
+        of `frequency`, then those of `l` and `m` (which are broadcast with
+        each other), and finally the row and column for matrices if
+        `output_type` is one of the matrix types.
```

Comment:
Feedback on the axis order is welcome. katholog works with the matrix indices as the leading (slowest-varying) axes, but my feeling is that for computation (as opposed to visualisation) one will need to work with the whole matrix rather than individual terms.